### DISCUSSION TITLE
How to install scala-cli  without internet

### DISCUSSION THREAD

**Discussion by halloleo** [VirtusLab/scala-cli#2869] [General]
I need to install scala-cli on a Linux RHEL machine without internet.

I have downloaded `cliBootstrapped-1.2.2.jar` and moved it to that machine - but I don't have Coursier on that system.

On that machine I have Java (versions 8 and 17) and Scala (version 3.3.1) installed. Is this enough to run scala-cli? A simple `java -jar cliBootstrapped-1.2.2.jar` doesn't seem to return.  (It might try to download something.)

How can I use `cliBootstrapped-1.2.2.jar` to run/install scala-cli?

**Comment by Gedochao** [comment]
`cliBootstrapped-x.y.z.jar` are self-executable (bootstrapped) jars.
You don't need the `java -jar` command, it can self execute on a Linux machine.
```bash
chmod +x cliBootstrapped-1.2.2.jar 
./cliBootstrapped-1.2.2.jar 
# Welcome to Scala 3.4.1 (17, Java OpenJDK 64-Bit Server VM).
# Type in expressions for evaluation. Or try :help.
#                                                                                                                  
# scala> 
```

the `--offline` flag will be your friend, as this is an experimental option which makes sure Scala CLI relies on its caches (rather than anything it gets from the net).
you normally would also need Bloop, but you can skip using (and downloading) Bloop with `--server=false`.

```bash
scala-cli -e 'println("Hello from offline mode")' --offline --server=false --power
# The `--offline` option is experimental
# Please bear in mind that non-ideal user experience should be expected.
# If you encounter any bugs or have feedback to share, make sure to reach out to the maintenance team at https://github.com/VirtusLab/scala-cli
# Hello from offline mode
```

What you will need, however, is prefilling your `coursier` cache with any dependencies you may want to use.
We don't really have a sub-command for prefilling caches, at this point it's a power-user thing.
What I would advise doing is running a Scala CLI command with any project dependencies you may have on a different machine with internet connection, and then copying the `coursier` caches over to the isolated machine you are setting up on.

Docs on `coursier` cache are here:
https://get-coursier.io/docs/cache

Alternatively prefilling your local `ivy2` with any data you may need should also work.
If your cache contains all you need (Scala compiler, jdk, any deps of your project, potentially Bloop if you want to skip `--server=false`), running things with `--offline` should get you set.

In case of problems, let me know what doesn't work.

**Reply by halloleo** [reply]
Thanks for the detailed reply, @Gedochao.

I have made `cliBootstrapped-1.2.2.jar` executable, but when I issue

    $ ./cliBootstrapped-1.2.2.jar

the command doesn't produce any output and never returns.

I also tried

    $ ./cliBootstrapped-1.2.2.jar --help

and 

    $ ./cliBootstrapped-1.2.2.jar --offline --help

same result: no output and never returns (so I guess it tries to download _something_ from the internet).

What do I do wrong? I don't use any dependencies and `scala`, `scalac` + `java` are on my path.

**Reply by Gedochao** [reply]
This is curious. No output and hangs?Completely nothing?
huh. that is something new.

We can also try `--server=false`, to eliminate Bloop out of the picture, in case that's what hangs...
```sh
././cliBootstrapped-1.3.1.jar --server=false
Welcome to Scala 3.4.1 (17, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                 
scala> 
```
or add extra logging with `-v -v -v`
```sh
./cliBootstrapped-1.3.1.jar --server=false -v -v -v
Fetching List(Dependency(org.scala-lang:scala3-compiler_3, 3.4.1, Configuration(), Set(), Publication(, Type(), Extension(), Classifier()), false, true)), adding List(IvyRepository(Pattern(List(Const(file:///Users/pchabelski/Library/Caches/ScalaCli/local-repo/1.3.1//), Var(organisation), Const(/), Var(module), Const(/), Opt(List(Const(scala_), Var(scalaVersion), Const(/))), Opt(List(Const(sbt_), Var(sbtVersion), Const(/))), Var(revision), Const(/), Var(type), Const(s/), Var(artifact), Opt(List(Const(-), Var(classifier))), Const(.), Var(ext))), None, None, true, true, true, false, None, true))
Found 12 artifacts:
(...)
-classpath
/Users/pchabelski/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.4.1/scala3-library_3-3.4.1.jar:/Users/pchabelski/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar

Welcome to Scala 3.4.1 (17, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                 
scala> 
```
Otherwise, there's not much to go on. 😕

**Reply by halloleo** [reply]
Thx @Gedochao. I will try --server=false and the verbose output when I'm back on the air-gapped machine (but this will be not before June).

**Reply by halloleo** [reply]
Ok, I'm back at the air-gapped machine. 

I am kicking off

    ./cliBootstrapped-1.2.2.jar --server-false -v -v -v

and the command runs for one minute or so and then returns with:

> Error: java.lang.Exception: java.lang.Exception: Error while getting https://github.com/coursier/jvm-index/raw/master/index.json: download error: Caught java.net.ConnectException (Connection timed out (Connection timed out)) while downloading https://github.com/coursier/jvm-index/raw/master/index.json

Why does this happen when I give `--server=false`?

**Reply by bishabosha** [reply]
you need the `--offline` and `--power` flags to be offline

**Reply by halloleo** [reply]
Thanks for this comment, @bishabosha. I have tried it:

    $ ./cliBootstrapped-1.2.2.jar --offline --power

but  scala-cli _still_ wants to download stuff:

> Error: java.lang.Exception: java.lang.Exception: Error while getting https://github.com/coursier/jvm-index/raw/master/index.json: download error: Caught java.net.ConnectException (Connection timed out (Connection timed out)) while downloading https://github.com/coursier/jvm-index/raw/master/index.json

Same when I do

    $ ./cliBootstrapped-1.2.2.jar --offline --power --server=false

**Reply by bishabosha** [reply]
Does it also happen if you use an explicit java command and made sure to set JAVA_HOME environment variable

**Reply by halloleo** [reply]
Ok, this pointed me in the right direction, thank you: The `JAVA_HOME` is normally set to a Java 8 JDK, but when I set it to Java 17 the result makes more sense:

> The `--offline` option is experimental
Please bear in mind that non-ideal user experience should be expected.
If you encounter any bugs or have feedback to share, make sure to reach out to the maintenance team at https://github.com/VirtusLab/scala-cli
[error]  Error downloading org.scala-lang:scala3-compiler_3:3.4.1
  not found: /home/me/.ivy2/local/org.scala-lang/scala3-compiler_3/3.4.1/ivys/ivy.xml
  not found: /home/me/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/3.4.1/scala3-compiler_3-3.4.1.pom
  not found: /home/me/.cache/scalacli/local-repo/1.2.2/org.scala-lang/scala3-compiler_3/3.4.1/ivys/ivy.xml
  No fallback URL found

I guess I have to download these three files on another machine and then put them on the air-gapped machine in the listed locations... 

Can I tell scala-cli to look in _one_ directory I can put these three files in?

**Reply by Gedochao** [reply]
The default way to do it would be to prefill your Coursier cache:
https://get-coursier.io/docs/cache
We use Coursier for finding dependencies.

You can override where the cache should be with the `COURSIER_CACHE` env variable (more details can be found in the Coursier docs).

You can simply use the Coursier `fetch` sub-command on another machine and copy the cache from there.
https://get-coursier.io/docs/cli-fetch

So in your case, smth like this would do the trick:
```bash
cs fetch org.scala-lang:scala3-compiler_3:3.4.1
```

**Reply by Gedochao** [reply]
Note that you will likely need all of the Scala 3.4.1 artifacts we use, so it might also be a good idea to install it whole with `install` (https://get-coursier.io/docs/cli-install):
```bash
cs install scala:3.4.1
```
and also the `scala3-sbt-bridge` if you're running Bloop (not necessary with `--server=false`):
```bash
cs fetch org.scala-lang:scala3-sbt-bridge:3.4.1
```

**Reply by halloleo** [reply]
I will try this. Looks very promising!

**Reply by halloleo** [reply]
It works now! Installing coursier on an internet-connected machine and then copying the coursier cache over to the air-gapped machine (where no coursier is installed) does the trick!

Thanks a lot for patiently sticking with my issue!
